### ISSUE TITLE
Wonderful work!

### ISSUE THREAD

**Issue by seekever** [microsoft/TRELLIS.2#21]
An impressive piece of work! I found another similar voxel representation called Faithful Contouring, and I'm curious about which one is more effective.

**Comment by seekever** [comment]
In addition, currently only TRELLIS.2 has released the VAE and image-to-3D models.

**Comment by JeffreyXiang** [comment]
Thanks for the reference and initiating this discussion! I believe both methods are inspired by the original Dual Contouring algorithm. Ours is a relatively simpler variant in terms of implementation and feature channel numbers, also it allows for flexible quad splitting optimized by rendering supervision. Additionally, we have already shown promising image to 3D generation results.

**Comment by seekever** [comment]
> Thanks for your reply. After testing your amazing 3D generation pipeline, I found that face flips often occur if the Dual Contouring remeshing is not applied. However, if remeshing is performed, double-layer shells occur due to UDF. In my tests, Faithful Contouring can produce a thin mesh without face flips. Is there a way for O-voxel to solve the face flip issue while maintaining a thin mesh?

**Comment by JeffreyXiang** [comment]
We do not include explicit normal information in our geometry representation, so the triangles should be treated as double-sided. Instead, normal information is stored with the textural data via `textured_mesh_to_volumetric_attr`, which accounts for the effects of normal maps.

If you need consistent face orientation, here are two potential solutions:
1.  **Geometric reorientation:** Without additional normal information, you can use algorithms like `unify_face_orientations` from `cumesh` (or similar libraries) to reorient the face directions.
2.  **Alignment with implicit normals:** You could flip triangles that are inconsistent with the normal vectors stored in `o_voxel`. Note that this would require a simple extension, as the current TRELLIS.2 model does not explicitly model a normal channel for the sake of simplicity.

**Comment by seekever** [comment]
> We do not include explicit normal information in our geometry representation, so the triangles should be treated as double-sided. Instead, normal information is stored with the textural data via `textured_mesh_to_volumetric_attr`, which accounts for the effects of normal maps.
> 
> If you need consistent face orientation, here are two potential solutions:
> 
> 1. **Geometric reorientation:** Without additional normal information, you can use algorithms like `unify_face_orientations` from `cumesh` (or similar libraries) to reorient the face directions.
> 2. **Alignment with implicit normals:** You could flip triangles that are inconsistent with the normal vectors stored in `o_voxel`. Note that this would require a simple extension, as the current TRELLIS.2 model does not explicitly model a normal channel for the sake of simplicity.

Good suggestions! As for unify_face_orientation, it has been included in your code.  https://github.com/microsoft/TRELLIS.2/blob/1762f493fe7731a3b7cc6b79ad5da7b015b516c1/o-voxel/o_voxel/postprocess.py#L162
I have tested it (setting remesh=False), but the face flips still exist. 
Alignment with implicit normals should work better.

**Comment by wendashi** [comment]
> We do not include explicit normal information in our geometry representation, so the triangles should be treated as double-sided. Instead, normal information is stored with the textural data via `textured_mesh_to_volumetric_attr`, which accounts for the effects of normal maps.
> 
> If you need consistent face orientation, here are two potential solutions:
> 
> 1. **Geometric reorientation:** Without additional normal information, you can use algorithms like `unify_face_orientations` from `cumesh` (or similar libraries) to reorient the face directions.
> 2. **Alignment with implicit normals:** You could flip triangles that are inconsistent with the normal vectors stored in `o_voxel`. Note that this would require a simple extension, as the current TRELLIS.2 model does not explicitly model a normal channel for the sake of simplicity.

We found that a directed edge extension of FDG (FDG -> FDG-D) can store the face orientation, and adding an adapter on shape vae can predict this information without compromising the pre-trained prior, and more details can be found at the paper AnySurf( https://arxiv.org/abs/2605.26149)

**Comment by xuxu116** [comment]
> > We do not include explicit normal information in our geometry representation, so the triangles should be treated as double-sided. Instead, normal information is stored with the textural data via `textured_mesh_to_volumetric_attr`, which accounts for the effects of normal maps.
> > If you need consistent face orientation, here are two potential solutions:
> > 
> > 1. **Geometric reorientation:** Without additional normal information, you can use algorithms like `unify_face_orientations` from `cumesh` (or similar libraries) to reorient the face directions.
> > 2. **Alignment with implicit normals:** You could flip triangles that are inconsistent with the normal vectors stored in `o_voxel`. Note that this would require a simple extension, as the current TRELLIS.2 model does not explicitly model a normal channel for the sake of simplicity.
> 
> We found that a directed edge extension of FDG (FDG -> FDG-D) can store the face orientation, and adding an adapter on shape vae can predict this information without compromising the pre-trained prior, and more details can be found at the paper AnySurf( https://arxiv.org/abs/2605.26149)

Nice idea, are you going to opensorce your FDG-D version of ovoxel？

**Comment by wendashi** [comment]
> > > We do not include explicit normal information in our geometry representation, so the triangles should be treated as double-sided. Instead, normal information is stored with the textural data via `textured_mesh_to_volumetric_attr`, which accounts for the effects of normal maps.
> > > If you need consistent face orientation, here are two potential solutions:
> > > 
> > > 1. **Geometric reorientation:** Without additional normal information, you can use algorithms like `unify_face_orientations` from `cumesh` (or similar libraries) to reorient the face directions.
> > > 2. **Alignment with implicit normals:** You could flip triangles that are inconsistent with the normal vectors stored in `o_voxel`. Note that this would require a simple extension, as the current TRELLIS.2 model does not explicitly model a normal channel for the sake of simplicity.
> > 
> > 
> > We found that a directed edge extension of FDG (FDG -> FDG-D) can store the face orientation, and adding an adapter on shape vae can predict this information without compromising the pre-trained prior, and more details can be found at the paper AnySurf( https://arxiv.org/abs/2605.26149)
> 
> Nice idea, are you going to opensorce your FDG-D version of ovoxel？

I'll open-source it once it's accepted😢And I can update the link here when it's released. Thanks for your attention.

**Comment by seekever** [comment]
> > We do not include explicit normal information in our geometry representation, so the triangles should be treated as double-sided. Instead, normal information is stored with the textural data via `textured_mesh_to_volumetric_attr`, which accounts for the effects of normal maps.
> > If you need consistent face orientation, here are two potential solutions:
> > 
> > 1. **Geometric reorientation:** Without additional normal information, you can use algorithms like `unify_face_orientations` from `cumesh` (or similar libraries) to reorient the face directions.
> > 2. **Alignment with implicit normals:** You could flip triangles that are inconsistent with the normal vectors stored in `o_voxel`. Note that this would require a simple extension, as the current TRELLIS.2 model does not explicitly model a normal channel for the sake of simplicity.
> 
> We found that a directed edge extension of FDG (FDG -> FDG-D) can store the face orientation, and adding an adapter on shape vae can predict this information without compromising the pre-trained prior, and more details can be found at the paper AnySurf( https://arxiv.org/abs/2605.26149)

Based on the analysis, the three axes in Faithful Contouring's edge_flux_sign that correspond to O-voxel's intersected can be identified. If we preserve the corresponding signs and make minor modifications to how flexible_dual_grid_to_mesh computes face orientation, correct face orientations can be obtained. Not sure how large the difference is between your FDG-D and Faithful Contouring.

**Comment by wendashi** [comment]
FC is more complex and not as minimal. More importantly, FC only implements it for VAE reconstruction, and whether it can be used for DiT generation remains to be verified.

<img width="2498" height="518" alt="Image" src="https://github.com/user-attachments/assets/143a713b-01c1-4caa-957c-61b1e6be197c" />

**Comment by seekever** [comment]
> FC is more complex and not as minimal. More importantly, FC only implements it for VAE reconstruction, and whether it can be used for DiT generation remains to be verified.
> 
> <img alt="Image" width="2000" height="415" src="https://private-user-images.githubusercontent.com/141135497/638839595-143a713b-01c1-4caa-957c-61b1e6be197c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODcyMzIzNTYsIm5iZiI6MTc4NzIzMjA1NiwicGF0aCI6Ii8xNDExMzU0OTcvNjM4ODM5NTk1LTE0M2E3MTNiLTAxYzEtNGNhYS05NTdjLTYxYjFlNmJlMTk3Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwODIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDgyMFQxMzIwNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzM1YTJhMTg5ZGJiMTUxMzIxOWYyODViNTg5NzU3ODY0OGJkOWJhOGViYzA2ZjBlMDUzZjMwYmE4ODRmZjc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.aH5mqTNhhDFnFJQlBt25uRQY8S2zZlyrImMYGmTVhWI">

The FCT V1.5 is simple

<img width="556" height="267" alt="Image" src="https://github.com/user-attachments/assets/6baafbf0-853b-4270-8b19-165eb6535f2b" />

Your 3 axis-aligned directed-edge maybe found in edge_flux_sign.

I understand that in the FCT1.5 decoder, face orientation is directly determined by edge_flux_sign; the normals are only used in the decode step to choose the diagonal split, and do not directly flip the faces.

Also, edge_flux_sign is redundant — neighboring voxels share edges, and 3 of them correspond to the ovoxel's intersected flag, i.e., it's like a directed intersected flag.

FCT indeed hasn't released the VAE or DiT. If you could release the FGD-D VAE and DiT, that would be a really significant contribution — looking forward to it!

**Comment by xuxu116** [comment]
> > FC is more complex and not as minimal. More importantly, FC only implements it for VAE reconstruction, and whether it can be used for DiT generation remains to be verified.
> > <img alt="Image" width="2000" height="415" src="https://private-user-images.githubusercontent.com/141135497/638839595-143a713b-01c1-4caa-957c-61b1e6be197c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODcyMzIzNTYsIm5iZiI6MTc4NzIzMjA1NiwicGF0aCI6Ii8xNDExMzU0OTcvNjM4ODM5NTk1LTE0M2E3MTNiLTAxYzEtNGNhYS05NTdjLTYxYjFlNmJlMTk3Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwODIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDgyMFQxMzIwNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzM1YTJhMTg5ZGJiMTUxMzIxOWYyODViNTg5NzU3ODY0OGJkOWJhOGViYzA2ZjBlMDUzZjMwYmE4ODRmZjc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.aH5mqTNhhDFnFJQlBt25uRQY8S2zZlyrImMYGmTVhWI">
> 
> The FCT V1.5 is simple
> 
> <img alt="Image" width="556" height="267" src="https://private-user-images.githubusercontent.com/5241948/638867317-6baafbf0-853b-4270-8b19-165eb6535f2b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODc3MzkyNjksIm5iZiI6MTc4NzczODk2OSwicGF0aCI6Ii81MjQxOTQ4LzYzODg2NzMxNy02YmFhZmJmMC04NTNiLTQyNzAtOGIxOS0xNjVlYjY1MzVmMmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDgyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA4MjZUMTAwOTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2M0NjZhZjBjYmE1ODk4NTk0Yzg3MTg0NGFiYzAwOGM2NzNiM2ZmMWJhZDM3ZDEyNWNkZGVjYTU5ZmU0YmFjMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.QKgDzsUUnkUBiZJd_3ZZ6zzfBcroEf1AC59Ko0ARYoY">
> Your 3 axis-aligned directed-edge maybe found in edge_flux_sign.
> 
> I understand that in the FCT1.5 decoder, face orientation is directly determined by edge_flux_sign; the normals are only used in the decode step to choose the diagonal split, and do not directly flip the faces.
> 
> Also, edge_flux_sign is redundant — neighboring voxels share edges, and 3 of them correspond to the ovoxel's intersected flag, i.e., it's like a directed intersected flag.
> 
> FCT indeed hasn't released the VAE or DiT. If you could release the FGD-D VAE and DiT, that would be a really significant contribution — looking forward to it!

Refer to the FDG-D, I have implmented the o-voxel with normal flag (https://github.com/xuxu116/o-voxel-with-normal). The VAE part is easy to train with additional input/output channels (similar to the origin intersected flags). If you don't want to change your flow model, it is fine to just finetune the Decoderonly and the normal_flag is fast to converge!
Below is the ovoxel-with-normal results:
<img width="4400" height="3351" alt="Image" src="https://github.com/user-attachments/assets/9b56d118-aad7-49f6-8100-3bdc2c2ea0ff" />

**Comment by wendashi** [comment]
As reported in our work (Table 6. (g) and (h) of https://arxiv.org/abs/2605.26149), to preserve the original model's pretrained prior as much as possible, adding an adapter branch instead of directly fine-tuning the shape VAE is a more practical approach.
